interpreter/CMakeLists.txt

```diff
@@ -134,24 +160,243 @@ set_directory_properties(PROPERTIES INCLUDE_DIRECTORIES "")
 set(LLVM_APPEND_VC_REV OFF CACHE BOOL "Disabled LLVM revision dependency" FORCE)
 
 #---Add the sub-directory excluding all the targets from all-----------------------------------------
-if(CMAKE_GENERATOR MATCHES "Xcode")
-  add_subdirectory(llvm/src)
+if(builtin_llvm)
+  if(CMAKE_GENERATOR MATCHES "Xcode")
+    add_subdirectory(llvm/src)
+  else()
+    add_subdirectory(llvm/src EXCLUDE_FROM_ALL)
+  endif()
+
+  #---Make LLVM compilation flags public---------------------------------------------------------------
+  # Extract the compilation flags from LLVM and make them public to the
+  # rest of ROOT so that we can compile against LLVM with matching flags.
+
+  # LLVM doesn't really give us a API to get this with an in-source build
+  # so we just use the normal way of doing this and read the llvm directory
+  # compilation properties.
+  get_directory_property(LLVM_DEFS DIRECTORY llvm/src COMPILE_DEFINITIONS)
+  # Turns DEFINE1;DEFINE2 to -DDEFINE1 -DDEFINE2
+  string (REPLACE ";" " -D" LLVM_DEFS ";${LLVM_DEFS}")
+  # CACHE variable that contains the additional flags that LLVM needs.
+  set(ROOT_LLVM_FLAGS "${LLVM_DEFS}" CACHE STRING "" FORCE)
 else()
-  add_subdirectory(llvm/src EXCLUDE_FROM_ALL)
-endif()
+  if (NOT builtin_clang)
+    message(FATAL_ERROR "ROOT cannot work with builtin clang. Please use -Dbuiltin_clang=Off to turn it off.")
+  endif()
+  # Rely on llvm-config.
+  set(CONFIG_OUTPUT)
+  find_program(LLVM_CONFIG NAMES "llvm-config-5.0" "llvm-config")
```

Comment:
Why not use LLVM's CMake modules? Macs don't have `llvm-config` from XCode, for example, although I understand that this is probably Linux-only since the version of LLVM has to almost match what's in ROOT.
On my machine, the file `LLVMConfig.cmake` has the flags returned by `llvm-config --cxxflags`, for example, as well as paths to various directories.

Reply:
This did not work well for llvm. One of the reasons is that clang would also include AddLLVM, which will include `LLVMConfig.cmake`.

My initial implementation was this but for some reason I had missing dependencies and dependency cycles. Then I decided to copy what clang's main CMakeLists.txt file is doing...

Reply:
What if we just add the clang subdirectory and let it handle how to find and configure LLVM for its needs? Would that work, maybe?

Reply:
It'd misplace the binaries. The copy-pasted code is not 1:1, it has a few things modified. Namely, it'd install 'private' cling/clang libs and binaries in `$ROOTSYS/bin/`. That would make things inconsistent if people build with `-Dbuiltin_llvm=On|Off`.

Reply:
I see. Ok, then.